connection/src/credentials.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+//! GRPC credentials support
+
+use grpcio::{Error as GrpcError, RpcStatusCode};
+use mc_util_grpc::BasicCredentials;
+use mc_util_uri::ConnectionUri;
+use std::{
+    fmt::{Debug, Display},
+    marker::PhantomData,
+};
+
+/// A trait that lets us determine if an error relates to authentication failure.
+pub trait AuthenticationError {
+    fn is_unauthenticated(&self) -> bool;
+}
+
+impl AuthenticationError for GrpcError {
+    fn is_unauthenticated(&self) -> bool {
+        match self {
+            GrpcError::RpcFailure(rpc_status) => {
+                rpc_status.status == RpcStatusCode::UNAUTHENTICATED
+            }
+            _ => false,
+        }
+    }
+}
+
+/// Error relating to credential providing.
+pub trait CredentialsProviderError: Debug + Display + Send + Sync {}
+impl<T> CredentialsProviderError for T where T: Debug + Display + Send + Sync {}
+
+/// An interface for providing credentials for a given URI.
+pub trait CredentialsProvider: Send + Sync {
+    type Uri: ConnectionUri;
+
+    fn get_credentials(
+        &self,
+        uri: &Self::Uri,
+    ) -> Result<Option<BasicCredentials>, Box<dyn CredentialsProviderError + 'static>>;
```

Comment:
> Any reason not to just use an associated type for the error?

I ran into issues with that with the `authenticatd_call` wrappers - I don't remember the specifics. It might work if the `CredentialsProvider` is a generic parameter of `ThickClient`.